### DISCUSSION TITLE
`spack install` and `spack upgrade` semantics

### DISCUSSION THREAD

**Discussion by tgamblin** [spack/spack#41882] [Ideas]
TL;DR: We're reaching the limits of how many jobs we entrust to a single `spack install` command and have it still make sense, so I'm proposing to simplify things by breaking it up into two commands (like many other systems do):
* `spack install`: your favorite spack command: tried and mostly true.
    * `spack install` should `--reuse` existing build caches and installations by default
    * It should *not* build anything if it doesn't have to and is optimized for consumers of packages.
* `spack upgrade`: a new command to bring existing installations up to date
    * This should freshen packages (but not their dependencies) by default
    * This is optimized for iterative workflows, developers, and other *builders* of packages.
    * This is the same workflow a regular package consumer wants if they end up wanting an upgrade, so it handles that case too (more intuitively than our current options).

## Goal

We want to make consumer and builder workflows easy.

## Motivation

I'll try to give a little history here to describe how we ended up where we are.

### 1. The beginning: fresh installs

Originally, every installation in Spack was an update, because we would do a fresh concretization for *everything*, without considering installed packages. If there was any reuse, it was because something concretized to exactly the same hash as a previous installation. 
* This is what we now call `spack install --fresh`.

### 2. Adding `--reuse`
Possibly everyone's biggest complaint about Spack for many years was the number of rebuilds needed to do any sort of installation. `--fresh` gets you the latest, and it's deterministic (for the same Spack commit) but it's slow and often not what users want. So:

* For `v0.17`, we added `--reuse` in #25310.
* For `v0.18`, we  made `--fresh` an option in `#28468` and reuse the default in #30396

### 3. The backlash: `reuse: dependencies`

Folks who were used to the olden days were not *entirely* pleased with the new semantics, as now by default if you said `spack install foo`, and `foo` was already installed, Spack would simply reuse the old version of `foo` and say it was installed. This is probably what *most* people want `install` to do, but for development workflows and for upgrades, it's not ideal.
* Developers who are iterating on a package really want it to rebuild, every time.
* `--fresh` is a blunt instrument: when you specify it, *nothing* is reused, and devs likely want their dependencies to stay put while iterating on some root package.
* The same is likely true for a normal user wanting to upgrade. `spack install --fresh` upgrades everything, including dependencies.

In response to this, we (probably too hastily):
* made `reuse:dependencies` the default in #30990.

This makes Spack reuse *only* dependencies but roots (here, `foo`) are fresh by default and `spack install foo` will reinstall whatever the latest `foo` is.  in an environment, `spack install` will always upgrade all the roots.

### 4. More issues: build caches and upstreams

Per the description of #41302, we've seen a lot of very confusing use cases where people have tried to push binaries to a build cache and have a friend install them directly from there. Without `--reuse`, this is a major pain, because you have to get your local configuration just right to ensure that specs from he cache are actually used.  This also makes it hard to use Spack in a pure binary workflow - i.e., where you have one dev producing binaries for a team, and *most* users are installing the produced binaries.

So now, we've:
* Proposed to put the default back the way it was in #41302
* Gotten some pushback from folks who really want a simplified iterative workflow (see examples in the PR).

## Splitting `install` and `upgrade`

I think the solution here is to split what is now `spack install` and a bunch of concretization options into `spack install` and `spack upgrade`. See the first section above for why.

## Specifics

There are a lot of cases to hear so I'm enumerating them for others' consideration. I'm hoping we can settle on something that allows both consumer workflows and developer/upgrade workflows to be intuitive.

### `install`

|                    | **No environment**                                                           | **In environment**                                                                  |
|------------------------------|------------------------------------------------------------------------------|-------------------------------------------------------------------------------------|
| `spack install`              | -                                                                            | install entire env, reuse all                                                       |
| `spack install --fresh`      | -                                                                            | install entire env, no reuse                                                        |
| `spack install SPEC`         | install SPEC, reusing roots *and* deps                                       | same, but SPEC must be added to env                                                 |
| `spack install --fresh SPEC` | install SPEC, all fresh, no reuse                                            | same, but SPEC must be added to env                                                 |

### `upgrade`

|                    | **No environment**                                                           | **In environment**                                                                  |
|------------------------------|------------------------------------------------------------------------------|-------------------------------------------------------------------------------------|
| `spack upgrade`              | -                                                                            | * `deconcretize` roots<br>* reconcretize all, reusing same deps<br>* install result |
| `spack upgrade --fresh`      | -                                                                            | * `deconcretize` roots<br>* reconcretize all, no reuse<br>* install result          |
| `spack upgrade SPEC`         | * reconcretize SPEC with fresh root<br>* also reuse deps<br>* install result | * deconcretize SPEC<br>* reconcretize SPEC reusing deps<br>* install resulting SPEC |
| `spack upgrade --fresh SPEC` | * reconcretize SPEC<br>* no reuse<br>* install result                        | * deconcretize SPEC<br>* reconcretize SPEC, no reuse<br>* install resulting SPEC    |

This is as far as I've gotten so would appreciate some feedback on the semantics.  Basically you can think of `spack upgrade` as `spack install`, but:

* The spec you pass it needs to already be installed (or at least concrete).
* The default concretization policy for `spack upgrade` is:
    * roots are fresh
    * dependencies are reused
* We `deconcretize` upgraded specs so that they're removed from an env's lockfile. This forces the thing you ask for to be upgraded *in an env*.

**Comment by tgamblin** [comment]
Would appreciate some feedback from @psakievich, @alecbcs, @becker33, @haampie, @alalazo, @adamjstewart, @michaelkuhn, @wdconinc, and anyone else who'd like to weigh in here.

**Reply by alecbcs** [reply]
This is great. I like all of the functionality outlined in this proposal and I think it'll go a long way to improving Spack's understand-ability for newer and less experienced users.

**Comment by alalazo** [comment]
When above we say:

> same, but SPEC must be added to env

do we mean that _it is already_ in the `spack.yaml` (and installation would fail otherwise)? Just trying to understand if we also plan to modify `--add` or not.

The other point is that, if we consider splitting the current `install` into these two commands, we can also consider having different defaults for them. For instance, I would prefer:

|                    | **No environment**                                                           | **In environment**                                                                  |
|------------------------------|------------------------------------------------------------------------------|-------------------------------------------------------------------------------------|
| `spack upgrade`              | -                                                                            | * `deconcretize` roots<br>* reconcretize all, no reuse<br>* install result |
| `spack upgrade --only-roots`      | -                                                                            | * `deconcretize` roots<br>* reconcretize all, reuse same deps<br>* install result          |
| `spack upgrade SPEC`         | * reconcretize SPEC with fresh root<br>*no reuse<br>* install result | * deconcretize SPEC<br>* reconcretize SPEC, no reuse<br>* install resulting SPEC |
| `spack upgrade --only-roots SPEC` | * reconcretize SPEC<br>* also reuse deps<br>* install result                        | * deconcretize SPEC<br>* reconcretize SPEC,  reuse same deps<br>* install resulting SPEC    |

The rationale is that, if a command is meant to update specs, I would find it weird if by default it does a very partial update, and needs further instructions to do a full update. Not sure this is a mainstream opinion though.

**Reply by tgamblin** [reply]
> do we mean that it is already in the spack.yaml (and installation would fail otherwise)?

Yep, not intending to modify that behavior — I think we *want* envs to require adding packages so that we don’t inadvertently modify `spack.yaml`.

> we can also consider having different defaults

So it looks like you want the default to *not* be to reuse deps, and `—only-roots` to be an option. Just going off experience from seeing teams here rebuild, and from seeing user complaints, what I’m trying to do is emphasize stability and make the defaults the more reliable options and require arguments for less stable things.

Basically: defaults should minimize builds to minimize risk. More specifically:

1. `install` will grab an existing install or binary and not build at all if it can. This is a main reason why I want `—reuse` to be default.
2. `upgrade` is an explicit ask for a change, but it defaults to upgrading *only* what you “asked for” and reusing what it can, again avoiding needless builds. In an env, you asked for roots and *don’t care* about deps (if you did, you’d specify them). On the CLI you said “upgrade this one spec specifically”. We should do that, again preferring binaries or existing installs by default if it can be accomplished that way.

I don’t think spack should do more than is necessary to accomplish what you asked. This translated to not building more packages than are required, because that introduces possibilities for build errors in packages you never cared about. I think it’s fairly intuitive that making `—fresh` default increases that risk.

The one place I would change my defaults above is “reuse existing deps”. Right now we need something between `-U` and `-fU`. In an env, the former keeps the lockfile (which is confusing) but won’t modify locked deps if it needs to. So it’s actually doing *less* than you asked for bc there are cases where it may not make necessary changes when it could. `-fU` throws the whole lockfile out and reconcretizes too much… I want something that finds the minimal changes to the lockfile (including dependencies), and I think that’s probably a better default for `upgrade`. It requires more work on the concretizer - basically we need to be able to make the lockfile a soft preference. I think this is beyond what I want to do in this discussion though.

**Reply by alecbcs** [reply]
> Right now we need something between -U and -fU.

+1. I'd prefer upgrade to be this hybrid `--fresh-roots` approach to minimize rebuilds by default with an option (`--fresh/--greedy`) to upgrade everything.

**Reply by alalazo** [reply]
> So it looks like you want the default to not be to reuse deps,

Not exactly. I think the semantic of `install` is good as is. 

Instead, I'd prefer if `spack update` defaulted to update everything, and selectively update only roots if explicitly asked via args. Rationale is that I'll ask for an update and by default update a small percentage of the nodes in the environment.

> In an env, you asked for roots and don’t care about deps (if you did, you’d specify them)

I don't agree 100% with this. Consider these 3 variations:
```yaml
spack:
  specs:
  - hdf5 ^openmpi

### 

spack:
  specs:
  - hdf5
  packages:
    mpi:
      require: openmpi

###

spack:
  specs:
  - hdf5
  - openmpi
```
According to what you say the latter cares about `openmpi` (and so `openmpi` will have a version update), while the first two don't (so `openmpi` will not be updated by default)? 

> I think it’s fairly intuitive that making —fresh default increases that risk.

I don't get this intuition... I think it depends on how old the environment is, and the deps. Probably combining fairly old deps with new roots causes failures too.

**Reply by alecbcs** [reply]
> Probably combining fairly old deps with new roots causes failures too.

We should be catching this in the `package.py` files through `depends_on` limits.

**Reply by alecbcs** [reply]
> According to what you say the latter cares about openmpi (and so openmpi will have a version update), while the first two don't (so openmpi will not be updated by default)?

This makes sense to me. As an application developer if I specify an environment as,
```yaml
spack:
  specs:
  - hdf5 ^openmpi
```

What I'm saying here is that I want `hdf5` that uses `openmpi` on the backend. I don't really care what version of `openmpi` is used as long as `hdf5` works/continues to work in subsequent environment upgrades.

**Reply by alalazo** [reply]
> We should be catching this in the package.py files through depends_on limits.

Sure, but all these are double way arguments. By the same argument, we should catch and fix build failures in `package.py` too, so having `--fresh` is not an issue. :slightly_smiling_face: 

To be clear, I'm not strongly opposed to any of the above, just gave my personal opinion on the default of `spack update` (which I think should be the opposite). 

The only note I have is that a few claims above seems more like other personal preferences. They are fine, but as I said I don't immediately see why having `--reuse-deps` is always minimizing build failures with respect to `--fresh`.

Or why this:
```yaml
spack:
  specs:
  - hdf5 ^openmpi
```
means I don't care about `openmpi` version while:
```yaml
spack:
  specs:
  - hdf5 
  - openmpi
```
means I do (to me all the environments above are kind of equivalent).

**Comment by alecbcs** [comment]
@tgamblin how do you imagine `upgrade`/`--fresh-roots` will handle newly deprecated dependency versions? Will we reuse an installed deprecated dependency version or reconcretize that dependency completely with reuse? (I'd vote for the latter for security.)

**Comment by adamjstewart** [comment]
So if I'm understanding correctly, you're proposing the following two changes:

1. `spack install` defaults to `--reuse`
2. `spack upgrade` becomes an alias for `spack install --reuse-deps`

If so, I'm fine with this behavior. It makes more sense linguistically. 

Some caveats to think about:

* We should compare with other PMs and make sure they have similar meaning for these commands
* #9134 got bogged down with `update` vs. `upgrade`, and this may have the same issue

**Reply by alecbcs** [reply]
Both `apt` (used by debian & ubuntu based distributions) and `brew` use a similar syntax,
- `update` - updates the package repository metadata. (Equivalent to a `git pull` for us.)
- `upgrade` - modify installed packages replacing with latest available.

Given both of those projects are extremely popular and often people's first introduction to a package manager, I'd be a big proponent of us adopting/reusing a similar syntax in Spack.

Also linking to #37549 which I think could be a nice complement to this `upgrade` proposal.

**Comment by psakievich** [comment]
I'm supportive of the concept.  I *think* I like the defaults @alalazo proposed of `spack upgrade` defaulting to no reuse, but I'm still mulling it over.  

I like the idea of basically one syntax being conservative (default of current `spack install`, makes sense to keep) and the other being aggressive by default (new `spack upgrade`).

There is definitely a large union between the two based on the proposed flags/options which I don't think is terrible.  However, spack's historical adoption of many syntaxes to accomplish the same thing means there is some analysis paralysis each time I spawn a new automated process rooted in spack.  

Polar opposite defaults make the intentions more clear, easier to communicate to users, and reduces the number of flags required to get what you want.
